// convai-sdk/dist/api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_promise_native_1 = __importDefault(require("request-promise-native"));
class ConvaiAPIClient {
    constructor(apiKey) {
        this.baseURL = "https://api.convai.dev/api/v1";
        this.apiKey = apiKey;
    }
    makeRequestWithBody(method, url, body) {
        return request_promise_native_1.default(`${this.baseURL}${url}`, {
            method,
            json: body,
            auth: {
                bearer: this.apiKey,
            }
        });
    }
    queryExecutions(query) {
        return this.makeRequestWithBody("POST", `/executions/query`, query);
    }
    trigger(req) {
        return this.makeRequestWithBody("POST", `/executions/trigger`, req);
    }
    broadcast(input) {
        return this.makeRequestWithBody("POST", "/executions/broadcast", input);
    }
    queryUsers(query) {
        return this.makeRequestWithBody("POST", `/users/super/query`, query);
    }
    queryUsersReachable(query) {
        return this.makeRequestWithBody("POST", "/users/super/query/reachable", query);
    }
    mergeUsers(req) {
        return this.makeRequestWithBody("POST", "/users/super/merge", req);
    }
    deleteSuperUser(id) {
        return this.makeRequestWithBody("DELETE", `/users/super/${id}`, null);
    }
    updateUserData(superUserId, input) {
        return this.makeRequestWithBody("PUT", `/users/super/${superUserId}`, input);
    }
    deleteChannelUser(userId) {
        return this.makeRequestWithBody("DELETE", `/users/channel/${userId}`, null);
    }
    updateSession(userId, input) {
        return this.makeRequestWithBody("PUT", `/users/session/${userId}`, input);
    }
    deleteSession(userId) {
        return this.makeRequestWithBody("DELETE", `/users/session/${userId}`, null);
    }
    setBaseURL(baseURL) {
        this.baseURL = baseURL;
    }
}
exports.ConvaiAPIClient = ConvaiAPIClient;
