// convai-sdk/dist/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const flaggable_1 = require("./flaggable");
const session_1 = require("./session");
class Context extends flaggable_1.Flaggable {
    constructor() {
        super();
        this.id = "";
        this.user = new RequestUser();
        this.session = new session_1.Session();
        this.text = "";
        this.channel = "";
        this.isStart = false;
        this.isTrigger = false;
    }
    /**
     * Converts from a plain js object into a Context class
     * @param obj an object with the same fields as a context
     */
    static fromObject(obj) {
        const c = new Context();
        c.id = obj.id;
        c.data = obj.data;
        c.user.data = obj.user.data;
        c.user.id = obj.user.id;
        c.user.channelId = obj.user.channelId;
        c.session.data = obj.session.data;
        c.session.ver = obj.session.ver;
        c.text = obj.text;
        c.channel = obj.channel;
        c.source = obj.source;
        c.isStart = obj.isStart;
        c.isTrigger = obj.isTrigger;
        c.errors = obj.errors;
        c.response = obj.response;
        return c;
    }
}
exports.Context = Context;
class RequestUser extends flaggable_1.Flaggable {
    constructor() {
        super();
        this.id = "";
        this.channelId = "";
    }
}
exports.RequestUser = RequestUser;
