// convai-sdk/dist/context-modifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ContextModifier {
    /**
     * Initialize the class
     */
    constructor() {
        this.changes = [];
        this.logs = [];
        this.errors = [];
    }
    /**
     * Add an operation for the Convai application
     *
     * @param type - The context modifier type
     * @param op - The type of operation
     * @param key - The key for the value
     * @param data - Value
     *
     * @returns The current context modifier instance
     */
    addOperation(type, op, key, data) {
        this.changes.push({
            type,
            op,
            key,
            data,
        });
        return this;
    }
    apply(ctx) {
        // TODO implement
    }
    error(err) {
        this.errors.push(err);
        return this;
    }
    /**
     * Add a key-value pair in the session scope
     *
     * @param key - The key for the value
     * @param data - The value to be added
     *
     * @returns The current context modifier instance with the newly added session key-value pair
     */
    setSession(key, data) {
        return this.addOperation(CMType.SESSION, CMOperation.SET, key, data);
    }
    /**
     * Add a key-value pair in the context/execution scope
     *
     * @param key - The key for the value
     * @param data - The value to be added
     *
     * @returns The current context modifier instance with the newly added context key-value pair
     */
    set(key, data) {
        return this.addOperation(CMType.CONTEXT, CMOperation.SET, key, data);
    }
    /**
     * Add a key-value pair in the user scope
     *
     * @param key - The key for the value
     * @param data - The value to be added
     *
     * @returns The current context modifier instance with the newly added user key-value pair
     */
    setUser(key, data) {
        return this.addOperation(CMType.USER, CMOperation.SET, key, data);
    }
    /**
     * Add a key-value pair in the environment scope
     *
     * @param key - The key for the value
     * @param data - The value to be added
     *
     * @returns The current context modifier instance with the newly added environment key-value pair
     */
    setEnvironment(key, data) {
        return this.addOperation(CMType.ENVIRONMENT, CMOperation.SET, key, data);
    }
    /**
     * Delete a key-value pair in the session scope
     *
     * @param key - The key for the value
     *
     * @returns The current context modifier instance with the deleted key-value pair
     */
    deleteSession(key) {
        return this.addOperation(CMType.SESSION, CMOperation.DELETE, key, null);
    }
    /**
     * Delete a key-value pair in the context/execution scope
     *
     * @param key - The key for the value
     *
     * @returns The current context modifier instance with the deleted key-value pair
     */
    delete(key) {
        return this.addOperation(CMType.CONTEXT, CMOperation.DELETE, key, null);
    }
    /**
     * Delete a key-value pair in the user scope
     *
     * @param key - The key for the value
     *
     * @returns The current context modifier instance with the deleted key-value pair
     */
    deleteUser(key) {
        return this.addOperation(CMType.USER, CMOperation.DELETE, key, null);
    }
    /**
     * Delete a key-value pair in the environment scope
     *
     * @param key - The key for the value
     *
     * @returns The current context modifier instance with the deleted key-value pair
     */
    deleteEnvironment(key) {
        return this.addOperation(CMType.ENVIRONMENT, CMOperation.DELETE, key, null);
    }
    /**
     * Clear all the key-value pairs in the context/execution scope
     *
     * @returns The current context modifier instance with the cleared key-value pairs
     */
    clear() {
        return this.addOperation(CMType.CONTEXT, CMOperation.CLEAR, "", null);
    }
    /**
     * Clear all the key-value pairs in the session scope'
     *
     * @returns The current context modifier instance with the cleared key-value pairs
     */
    clearSession() {
        return this.addOperation(CMType.SESSION, CMOperation.CLEAR, "", null);
    }
    /**
     * Clear all the key-value pairs in the user scope
     *
     * @returns The current context modifier instance with the cleared key-value pairs
     */
    clearUser() {
        return this.addOperation(CMType.USER, CMOperation.CLEAR, "", null);
    }
    /**
     * Clear all the key-value pairs in the environment scope
     *
     * @returns The current context modifier instance with the cleared key-value pairs
     */
    clearEnvironment() {
        return this.addOperation(CMType.ENVIRONMENT, CMOperation.CLEAR, "", null);
    }
    /**
     * Add a log message be sent with the API request
     *
     * @param level - Log level for proper coloring
     * @param message - Log message
     *
     * @returns The current context modifier instance
     */
    log(level, message) {
        this.logs.push({
            level,
            message,
            time: new Date(),
        });
        return this;
    }
    /**
     * Add a trace level log
     *
     * @param message - Message for the log
     *
     * @returns The current context modifier instance with the added log
     */
    logTrace(message) {
        return this.log(LogLevel.TRACE, message);
    }
    /**
     * Add a debug level log
     *
     * @param message - Message for the log
     *
     * @returns The current context modifier instance with the added log
     */
    logDebug(message) {
        return this.log(LogLevel.DEBUG, message);
    }
    /**
     * Add a info level log
     *
     * @param message - Message for the log
     *
     * @returns The current context modifier instance with the added log
     */
    logInfo(message) {
        return this.log(LogLevel.INFO, message);
    }
    /**
     * Add a warning level log
     *
     * @param message - Message for the log
     *
     * @returns The current context modifier instance with the added log
     */
    logWarning(message) {
        return this.log(LogLevel.WARNING, message);
    }
    /**
     * Add a error level log
     *
     * @param message - Message for the log
     *
     * @returns The current context modifier instance with the added log
     */
    logError(message) {
        return this.log(LogLevel.ERROR, message);
    }
}
exports.ContextModifier = ContextModifier;
var CMType;
(function (CMType) {
    CMType[CMType["CONTEXT"] = 0] = "CONTEXT";
    CMType[CMType["SESSION"] = 1] = "SESSION";
    CMType[CMType["USER"] = 2] = "USER";
    CMType[CMType["ENVIRONMENT"] = 3] = "ENVIRONMENT";
})(CMType || (CMType = {}));
var CMOperation;
(function (CMOperation) {
    CMOperation[CMOperation["SET"] = 0] = "SET";
    CMOperation[CMOperation["DELETE"] = 1] = "DELETE";
    CMOperation[CMOperation["CLEAR"] = 2] = "CLEAR";
})(CMOperation || (CMOperation = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["TRACE"] = 0] = "TRACE";
    LogLevel[LogLevel["DEBUG"] = 5] = "DEBUG";
    LogLevel[LogLevel["INFO"] = 10] = "INFO";
    LogLevel[LogLevel["WARNING"] = 15] = "WARNING";
    LogLevel[LogLevel["ERROR"] = 20] = "ERROR";
})(LogLevel || (LogLevel = {}));
