// convai-sdk/dist/user-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class UserQueryBuilder {
    /**
     * Initialize the class
     *
     * @param mode Query Mode to use for the builder. Can be ANY | ALL | NONE
     * @param limit Limit the number of results
     * @param offset Offset the results
     */
    constructor(mode, limit) {
        this.mode = mode || UserQueryMode.ALL;
        this.checks = [];
        this.limit = limit || 10;
        this.offset = 0;
    }
    /**
     * Set the offset on results
     *
     * @param offset - The offset to be set
     */
    setOffset(offset) {
        this.offset = offset;
        return this;
    }
    /**
     * Intialize a query to be performed. This method handles the 'key' part of the query
     *
     * @param field Name of the key
     */
    where(field) {
        // Check if a query is already in place. If yes, add it to the checks variable
        if (this.currentCheck) {
            this
                .checks
                .push(this.currentCheck);
        }
        this.currentCheck = {
            field,
            operation: UserQueryOperation.EQUALS,
            values: []
        };
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the return set from where 'key' is equal to 'value'
     *
     * @param values The value to be compared
     */
    equals(...values) {
        // Use of this method requires a where() method
        if (!this.currentCheck) {
            throw new Error("cannot call equals without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.EQUALS;
        this.currentCheck.values = values;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the return set from where 'key' doesn't equal 'value'
     *
     * @param values The value to be compared
     */
    notEquals(...values) {
        if (!this.currentCheck) {
            throw new Error("cannot call notEquals without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.NOT_EQUALS;
        this.currentCheck.values = values;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the return set from the where 'key' starts with 'value'
     *
     * @param values The value to be compared
     */
    startsWith(...values) {
        if (!this.currentCheck) {
            throw new Error("cannot call startsWith without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.STARTS_WITH;
        this.currentCheck.values = values;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the return set from the where 'key' should be greater than 'value'
     *
     * @param values The value to be compared
     */
    greaterThan(...values) {
        if (!this.currentCheck) {
            throw new Error("cannot call notEquals without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.NOT_EQUALS;
        this.currentCheck.values = values;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the return set from the where 'key' should be less than 'value'
     *
     * @param values The value to be compared
     */
    lessThan(...values) {
        if (!this.currentCheck) {
            throw new Error("cannot call lessThan without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.LESS_THAN;
        this.currentCheck.values = values;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the user data should have the where 'key'
     */
    exists() {
        if (!this.currentCheck) {
            throw new Error("cannot call exists without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.EXISTS;
        return this;
    }
    /**
     * This method handles the 'value' part of the query. Denotes that the user data should not have the where 'key'
     */
    notExists() {
        if (!this.currentCheck) {
            throw new Error("cannot call notExists without calling where first");
        }
        this.currentCheck.operation = UserQueryOperation.NOT_EXISTS;
        return this;
    }
    /**
     * Build different user queries into a single user query
     */
    build() {
        if (this.currentCheck) {
            this
                .checks
                .push(this.currentCheck);
        }
        return { mode: this.mode, checks: this.checks, limit: this.limit, offset: this.offset };
    }
}
exports.UserQueryBuilder = UserQueryBuilder;
var UserQueryOperation;
(function (UserQueryOperation) {
    UserQueryOperation[UserQueryOperation["EQUALS"] = 0] = "EQUALS";
    UserQueryOperation[UserQueryOperation["EXISTS"] = 1] = "EXISTS";
    UserQueryOperation[UserQueryOperation["NOT_EXISTS"] = 2] = "NOT_EXISTS";
    UserQueryOperation[UserQueryOperation["NOT_EQUALS"] = 3] = "NOT_EQUALS";
    UserQueryOperation[UserQueryOperation["STARTS_WITH"] = 4] = "STARTS_WITH";
    UserQueryOperation[UserQueryOperation["GREATER_THAN"] = 5] = "GREATER_THAN";
    UserQueryOperation[UserQueryOperation["LESS_THAN"] = 6] = "LESS_THAN";
})(UserQueryOperation = exports.UserQueryOperation || (exports.UserQueryOperation = {}));
var UserQueryMode;
(function (UserQueryMode) {
    UserQueryMode[UserQueryMode["ANY"] = 0] = "ANY";
    UserQueryMode[UserQueryMode["ALL"] = 1] = "ALL";
    UserQueryMode[UserQueryMode["NONE"] = 2] = "NONE";
})(UserQueryMode = exports.UserQueryMode || (exports.UserQueryMode = {}));
